### PR TITLE
Allow an existing secret to be used to get the auth token out of the values file

### PR DESCRIPTION
**Pull request by wreed4**
Without this change, it is impossible to install this chart without placing a secret value in the values file.  This has problem for installing with helm, but it has more problems for committing these values files to github.  Allowing users to use their own secret allows them to decide how the secret is stored and how it gets onto the cluster.

### CODE REVIEW

**Review by awly — APPROVED** [review]
> Without this change, it is impossible to install this chart without placing a secret value in the values file.

_Technically_, you could pass the token via a flag instead of the values file like so:
```
helm install teleport-kube-agent . \¬
   --set authToken=$(kubectl get teleport-kube-agent-join-token)
```

But it feels a bit awkward, so I'm OK with this PR.

**Comment by klizhentas** [comment]
Let me think about this a bit more before merging it. We can of course add more flags to support different use cases, but I would like look for alternative solutions.

**Review by klizhentas — CHANGES_REQUESTED** [review]
Thank you for your submission. Let me think about this a bit more before merging it. We can of course add more flags to support different use cases, but I would like look for alternative solutions.

**Comment by webvictim** [comment]
I feel like Helm should be able to check whether the secret exists before creating it. This could also tie into some work to allow the use of dynamic tokens.

**Comment by wreed4** [comment]
> Technically, you could pass the token via a flag instead of the values file like so:

We can't.  We manage helm charts declaratively via ArgoCD, not with `helm`.  So all configuration must be in github.  Also we will be installing this agent chart on 10s to 100s of clusters.  So it's infeasible to change them on individual clusters when we need to rotate them.  We need to pull them from a central spot with https://github.com/external-secrets/kubernetes-external-secrets .  

In addition, if the value is set in the `values` construct then it is stored and managed by helm as well as stored in a secret.  it is possible to grant users access to secrets created by helm (thereby allowing them to install/update helm charts) without granting them access to all secrets (to protect the information therein).  Storing the token in the `values` prevents that.

In general, any time a helm chart creates a secret, it seems standard practice to allow the installer to provide their own instead.  It's done in nearly all helm charts I've seen.  There are different patterns for doing it with different levels of nesting in the `values.yaml`, but as your file is flat, and helm best practices suggest to keep it as flat as possible, this seemed like a clean solution.

**Comment by wreed4** [comment]
>I feel like Helm should be able to check whether the secret exists before creating it.

I just learned today, it can!! https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-lookup-function

But this still doesn't work for our use-case and I assume many others.  you cannot guarantee that the secret will exist at the time of `helm template` and in fact, I can guarantee that it won't.  Our helm chart depends on yours and in our `templates` director we create an `ExternalSecret`.  So the secret will not exist until some time after the helm chart is deployed.  I believe this will be the case in nearly every enterprise usage of this chart since most cases require secrets get automatically pulled from somewhere secure like Vault or SSM.  We don't want secrets lying around on engineers laptops because they had to `kubectl create secret` manually before the chart existed.  Which wouldn't work anyways because as I mentioned before, we are using a high level of automation around the deployment of this component.

**Comment by klizhentas** [comment]
@wreed4 I think your use case is valid and probably the preferred one - using secrets vs passing tokens is better. I am thinking may be make this a default use case while deprecate the explicit token value.

**Comment by robcoward** [comment]
👍  for using  k8s `externalSecrets` to pull the token from an external location into a k8s secret.

File: examples/chart/teleport-kube-agent/values.yaml [hunk 1]
```diff
@@ -72,3 +72,5 @@ clusterRoleBindingName: teleport-kube-agent
 serviceAccountName: teleport-kube-agent
 # Name of the Secret to store the teleport join token.
 secretName: teleport-kube-agent-join-token
+# Use existing secret instead of creating one
```
Comment (wreed4) [inline examples/chart/teleport-kube-agent/values.yaml:75]:
```suggestion
# Use existing secret instead of creating one. the key must be `auth-token`.
```

**Comment by wreed4** [comment]
Hey guys, any chance of this getting merged soon? :) Anything else you need from me?

**Comment by ymatsiuk** [comment]
I think the same could be achieved with 
```yaml
{{- if not .Values.authToken }}
```
I see no need for extra parameter in `values.yaml`.
We might want to add this into `README.md`

```shell
export TELEPORT_KUBE_TOKEN=`<literal token> | base64 -w0`

cat <<EOF > secrets.yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: teleport-kube-agent-join-token
  namespace: production
type: Opaque
data:
  auth-token: ${TELEPORT_KUBE_TOKEN}
EOF

kubectl apply -f secret.yaml
```

**Comment by webvictim** [comment]
Hey @wreed4 - I'm going to close this in favour of https://github.com/gravitational/teleport/pull/6273 which was is little clearer and doesn't involve adding an extra flag. This should enable your use-case too - create the secret outside the chart, update the `secretName` value and leave the `authToken` value blank to use a custom secret.

**Pull request closed without merging**